src/main.rs

```diff
@@ -27,72 +29,29 @@ impl Default for Snapshots {
 
 impl Snapshots {
     pub fn new() -> Self {
-        let timestamp = SystemTime::now();
+        let snapshot = metriken_exposition::Snapshotter::default().snapshot();
 
-        let mut current = HashMap::new();
-
-        for metric in metriken::metrics().iter() {
-            let any = if let Some(any) = metric.as_any() {
-                any
-            } else {
-                continue;
-            };
-
-            let key = metric.name().to_string();
-
-            let snapshot = if let Some(histogram) = any.downcast_ref::<metriken::AtomicHistogram>()
-            {
-                histogram.snapshot()
-            } else if let Some(histogram) = any.downcast_ref::<metriken::RwLockHistogram>() {
-                histogram.snapshot()
-            } else {
-                None
-            };
-
-            if let Some(snapshot) = snapshot {
-                current.insert(key, snapshot);
-            }
-        }
-
-        let deltas = current.clone();
+        let previous: HistogramSnapshots = snapshot.histograms.into_iter().collect();
+        let delta = previous.clone();
 
         Self {
-            timestamp,
-            previous: current,
-            deltas,
+            timestamp: snapshot.systemtime,
+            previous,
+            delta,
         }
     }
 
     pub fn update(&mut self) {
-        self.timestamp = SystemTime::now();
-
-        let mut current = HashMap::new();
+        let mut snapshot = metriken_exposition::Snapshotter::default().snapshot();
+        self.timestamp = snapshot.systemtime;
 
-        for metric in metriken::metrics().iter() {
-            let any = if let Some(any) = metric.as_any() {
-                any
-            } else {
-                continue;
-            };
+        let mut current: HistogramSnapshots = snapshot.histograms.drain(..).collect();
+        let mut deltas: HistogramSnapshots = HashMap::new();
 
-            let key = metric.name().to_string();
-
-            let snapshot = if let Some(histogram) = any.downcast_ref::<metriken::AtomicHistogram>()
-            {
-                histogram.snapshot()
-            } else if let Some(histogram) = any.downcast_ref::<metriken::RwLockHistogram>() {
-                histogram.snapshot()
-            } else {
-                None
-            };
-
-            if let Some(snapshot) = snapshot {
-                if let Some(previous) = self.previous.get(&key) {
-                    self.deltas
-                        .insert(key.clone(), snapshot.wrapping_sub(previous).unwrap());
-                }
-
-                current.insert(key, snapshot);
+        for (name, previous) in &self.previous {
+            if let Some((name, current)) = current.remove_entry(name) {
```

Comment:
So I'm a little confused as to what is happening here: if we remove the entries, won't it be empty when `self.previous = current` happens on 58?

Reply:
TBH I'm not sure what my reasoning was since this PR is over a month old now. Fixed.